subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c

```diff
@@ -0,0 +1,743 @@
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_ACTIVATE);
+}
+
+static int deactivate_cb(uint16_t obj_inst_id)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_DEACTIVATE);
+}
+
+static int package_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+			    uint8_t *data, uint16_t data_len, bool last_block, size_t total_size)
+{
+	int ret = -EINVAL;
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	ret = handle_event(instance, EVENT_PKG_URI_WRITE);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = instance->write_package_cb(obj_inst_id, res_id, res_inst_id, data, data_len,
+					 last_block, total_size);
+
+	if (ret < 0) {
+		handle_event(instance, EVENT_DOWNLOAD_FAILED);
+		switch (ret) {
+		case -ENOMEM:
+			instance->update_result = UPD_RES_OUT_OF_MEMORY_DURING_DOWNLOAD;
+			break;
+		case -ENOSPC:
+			instance->update_result = UPD_RES_NOT_ENOUGH_STORAGE;
+			ret = -EFBIG;
+			break;
+		case -EFAULT:
+			instance->update_result = UPD_RES_PACKAGE_INTEGRITY_CHECK_FAILURE;
+			break;
+		default:
+			instance->update_result = UPD_RES_LOST_CONNECTION_DURING_DOWNLOAD;
+			break;
+		}
+
+		return ret;
+	}
+
+	if (last_block) {
+		handle_event(instance, EVENT_PKG_WRITTEN);
+	}
+
+	return 0;
+}
+
+static void set_update_result_from_error(struct firmware_pull_context *context, int error_code)
+{
+	struct lwm2m_swmgmt_data *instance;
+	instance = CONTAINER_OF(context, struct lwm2m_swmgmt_data, pull_context);
+
+
+	if (error_code == 0) {
+		handle_event(instance, EVENT_PKG_WRITTEN);
+
+		/* Do this somehow? */
+		handle_event(instance, EVENT_PKG_INTEGRITY_VERIFIED);
+		return;
+	}
+
+	handle_event(instance, EVENT_DOWNLOAD_FAILED);
+	if (error_code == -ENOMEM) {
+		instance->update_result = UPD_RES_OUT_OF_MEMORY_DURING_DOWNLOAD;
+	} else if (error_code == -ENOSPC) {
+		instance->update_result = UPD_RES_NOT_ENOUGH_STORAGE;
+	} else if (error_code == -EFAULT) {
+		instance->update_result = UPD_RES_PACKAGE_INTEGRITY_CHECK_FAILURE;
+	} else if (error_code == -ENOTSUP) {
+		instance->update_result = UPD_RES_INVALID_URI;
+	} else {
+		instance->update_result = UPD_RES_LOST_CONNECTION_DURING_DOWNLOAD;
+	}
+}
+
+static int package_uri_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+				uint8_t *data, uint16_t data_len, bool last_block,
+				size_t total_size)
+{
+#ifdef CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
+	int error_code;
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	instance->pull_context.write_cb = instance->write_package_cb;
+	instance->pull_context.result_cb = set_update_result_from_error;
+	instance->pull_context.obj_inst_id = instance->obj_inst_id;
+
+	strncpy(instance->pull_context.uri, instance->package_uri, PACKAGE_URI_LEN - 1);
+	error_code = lwm2m_pull_context_start_transfer(&instance->pull_context, K_NO_WAIT);
```

Comment:
Is the PR complete? I don't see this being defined anywhere.

Reply:
It's not. I've realized there are commits that this depends on and I'm working on integrating them into this PR. Sorry for the confusion.